Loki: Add documentation for app plugin developers and in code docs (#76067)
grafana/grafana: public/app/plugins/datasource/loki/APP_PLUGIN_DOCUMENTATION.md

# Loki data source documentation for app plugins developers

> Welcome to the developer documentation for the Loki data source! This documentation is designed to assist app plugin developers in leveraging the power of Loki data source within Grafana. Loki data source provides a set of APIs to simplify common tasks, making it easier to create engaging app plugins.

Reviewer: ```suggestion
Welcome to the developer documentation for the Loki data source! The purpose of this document is to assist app plugin developers in leveraging the power of the Loki data source and Grafana. The Loki data source provides a set of APIs to simplify common tasks, making it easier to create engaging app plugins.
```

Feel free to adjust this any any other suggestion.

## Introduction

> Loki data source provides a variety of methods, but not all of them are suitable for external use. In this documentation, we will focus on the key APIs that are highly recommended for app plugin development.

Reviewer: ```suggestion
The Loki data source provides a variety of methods, but not all of them are suitable for external use. In this documentation, we will focus on the key APIs that are highly recommended for app plugin development.
```

> It's important to note that the other APIs and methods that are not mentioned here may undergo changes or may not be suitable for external integration. Therefore, we do not recommend relying on them for your development needs.

Reviewer: ```suggestion
It's important to note some APIs and methods were deliberately ommited, as those may undergo changes or are not suitable for external integration. Therefore, we do not recommend relying on them for your development needs.
```

## Recommended APIs

> e strongly advise using these recommended APIs instead of direct proxy calls because they offer several benefits:

Reviewer: ```suggestion
We strongly advise using these recommended methods instead of direct API calls because they offer several benefits:
```

I think it would be positive to make a disticintion between server API and data source methods, so it's easier to know when you're speaking about an XHR request or a method call, to reduce ambiguity.

Even though that, I think it's also completely fine to refer to the set of public methods of the data source as "Loki data source API" or "data source API".

- Caching of results: These APIs intelligently cache results based on their input arguments and the currently selected time range. This caching mechanism improves performance by reducing redundant fetch requests.

- Proper authentication handling: The recommended APIs correctly handle custom authentication data source options, ensuring seamless authentication when making requests. This enhances both security and data access control.

> - Improved instrumentation: By opting for these recommended APIs, you gain access to advanced instrumentation features. These features provide better monitoring and analysis capabilities, offering deeper insights into your data source interactions.

Reviewer: How would an external developer use this instrumentation? Honest question, because I can't imagine.

Author: If app developer has an access to server logs, then they can see in more details why are their resource calls/requests failing, as those calls have better instrumentation (metrics/logs/traces). So for example, if some user says that they don't see available labels in app plugin, plugin developer can go do server logs and find logs and traces for when user was trying to fetch label names. And we have a significantly improved instrumentation for requests done trough `callResource` and `queryData`.

Reviewer: Can we add at least a brief mention on how to access it? If out out scope, no problem.

Author: I added a note https://github.com/grafana/grafana/pull/76067/commits/c82321a34a25cac358f2c33fe8e8ce82d2bb37ad

### Fetching Loki label names

> `datasource.languageProvider.fetchLabels()` can be used to fetch label names from connected Loki data source. Labels are essential for organizing and filtering log data. You can use this method to retrieve labels, providing a foundation for various data manipulation tasks within your app plugin.

Reviewer: Would it be possible to an external developer to work with custom LanguageProvider instances? Or should they always use it from a data source instance?

> `datasource.languageProvider.fetchLabels()` can be used to fetch label names from connected Loki data source. Labels are essential for organizing and filtering log data. You can use this method to retrieve labels, providing a foundation for various data manipulation tasks within your app plugin.

Reviewer: ```suggestion
`datasource.languageProvider.fetchLabels()` can be used to fetch label names from the connected Loki data source. Labels are essential for selecting and filtering log data. You can use this method to retrieve labels, providing a foundation for various data manipulation tasks within your app plugin.
```

/**
 * Example usage:
 */

const labelName = 'job';
try {
  const values = await fetchLabelValues(labelName);
  console.log(values);
} catch (error) {
  console.error(`Error fetching label values: ${error.message}`);
}
```

### Fetching Loki labels for a specified selector

`datasource.languageProvider.fetchSeriesLabels` can be used to fetch available labels for a given stream selector.

```ts
/**
 * Fetch series labels for a selector
 *
 * This method fetches labels for a given stream selector, such as `{job="grafana"}`.
 * It returns a promise that resolves to a record mapping label names to their corresponding values.
 *
 * @param streamSelector - The stream selector for which you want to retrieve labels.
 * @returns A promise containing a record of label names and their values.
 * @throws An error if the fetch operation fails.
 */
async function fetchSeriesLabels(streamSelector: string): Promise<Record<string, string[]>>;

/**
 * Example usage:
 */
const streamSelector = '{job="grafana"}';
try {
  const labels = await fetchSeriesLabels(streamSelector);
  console.log(labels);
} catch (error) {
  console.error(`Error fetching labels: ${error.message}`);
}
```

> ### Fetching Loki parser and label keys for log stream based on sampled lines

Reviewer: ```suggestion
### Detecting Loki parser and label keys for log stream based on sampled lines
```

I think it's important to use "Detecting" or "Inferring" because it's what the code does, unless I misremember.

> `datasource.languageProvider.getParserAndLabelKeys` is responsible for fetching parsers and label keys associated with a selected log stream. It achieves this by executing a small query to Loki with the specified stream and extracting relevant information from the received log lines.

Reviewer: ```suggestion
`datasource.languageProvider.getParserAndLabelKeys` receives a stream selector and returns the parsers and label keys that the query would generate. This is achieved by executing a sample query to Loki with the specified stream and extracting relevant information from the received log lines.
```
